smt: added fallback type

hackaugusto opened the pull request:
### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change
- [ ] Changelog entry
- [x] Used meaningful commit messages (should be good enough)

### Description

~~Fixes~~ Partial fix for #4538 

The problem in short: Unsupported variables were not added to `m_variables` but the visitor callbacks expected it to be there.

The solution: Always define the variables, if it's unsupported use a fallback type (uint256)

Opening the PR for code review, then I can add the tests and changelog (if that is okay)

hackaugusto committed:
smt: added fallback type

leonardoalt commented:
Hi @hackaugusto , thanks for the contribution!

I was wondering, does it also fix #4537 #4564 #4566 ?

axic commented:
@hackaugusto test failures seem to be irrelevant, but can you please rebase this?

axic commented on libsolidity/formal/SSAVariable.cpp:
```diff
@@ -39,7 +40,10 @@ SSAVariable::SSAVariable(
+		// Use a fallback type. This allow the SMT checker to run on functions
+		// which use unsupported types. It may introduce false-positivies, but
+		// it will allows the checker to run on the known types.
+		m_symbolicVar = make_shared<SymbolicFallbackVariable>(_decl, _interface);
```
Would it be worth issuing a warning here with a condensed version of the above text?

hackaugusto replied:
The warning is emitted in the [`createVariable` function](https://github.com/ethereum/solidity/pull/4949/files#diff-8ee88fd6f33965616a4a99fe36c2ba48R792)

hackaugusto commented:
> I was wondering, does it also fix #4537 #4564 #4566 ?

Not really

leonardoalt reviewed (changes requested):
@hackaugusto Could you add a test in `test/libsolidity/SMTChecker.cpp` with this fail?

leonardoalt commented:
I think when the test is added this PR is good to go.
Fallback expressions and assignments are still needed to make `MerkleProof.sol` stop crashing, but those should be added in different PRs.

axic commented:
> Fallback expressions and assignments are still needed to make MerkleProof.sol stop crashing

So this PR shouldn't close #4538 or at least the crash will be gone and will be replaced with an unsupported feature message?

leonardoalt commented:
There's still an assertion fail, though it's a different crash from #4538

hackaugusto committed:
Regression test for unsuported types

hackaugusto commented:
@leonardoalt added a regression test, there are a few drawbacks with this test though:

- It doesn't test the fallback type directly, it relies on `bytes32` not being supported, so, once support is added the test will not make much sense for the fallback type (but it should still be a useful as a regression test)
- It just asserts that the run is successful, without considering the warning messages, this could be useful to detect once support is added to `bytes32`, the test would have to be changed to a new type that is not supported. The problem is that eventually all types should be supported and either the feature would have to be removed or a new test type would be needed.
- [edit 2] The test is a minimal failing version of the original smart contract, which did not catch all bugs of the original.

edit: is a changelog entry required for this one?

hackaugusto committed:
Using int as a falblack type in the SMT solver

If an expression is found, which has an unsupported type, a integer
variable will be created to allow the solver to run.

hackaugusto committed:
added test for smt fallback type

axic commented on test/libsolidity/SMTChecker.cpp:
```diff
@@ -751,6 +751,24 @@ BOOST_AUTO_TEST_CASE(division_truncates_correctly)
 	)";
 	CHECK_SUCCESS_NO_WARNINGS(text);
 }
+BOOST_AUTO_TEST_CASE(regression_test)
```
I don't think regression test is a good name for this.

hackaugusto replied:
any suggestions?

leonardoalt replied:
I'd suggest `unsupported_types`

axic reviewed (changes requested):
Please rebase.

axic commented:
@leonardoalt apart from a rebase, are all the questions addressed here?

leonardoalt commented:
@axic I'm not sure what happens when rebasing this PR. I refactored how Solidity types are encoded into SMT sorts, so we'd have to see how the code fits together.

axic commented:
@hackaugusto @leonardoalt most of the relevant refactoring has been merged. It would be a good time checking this PR.

leonardoalt commented:
I think the current SMT refactor changes so many things that this PR would have to be rewritten. Also, the functionality given here is present in parts of the refactor.

axic commented:
Can you close it then?  It is a shame it wasn't addressed in time, but @hackaugusto your contribution is very much appreciated.

leonardoalt commented:
Yes, unfortunately closing seems the reasonable option.
